runtime/sema/check_reference_expression.go

```diff
@@ -34,9 +34,16 @@ func (checker *Checker) VisitReferenceExpression(referenceExpression *ast.Refere
 
 	var referenceType *ReferenceType
 	var targetType, referencedType Type
+	var returnType Type
 
 	if !resultType.IsInvalidType() {
 		var ok bool
+		// indexed access to dictionaries require an optional reference as their
+		// target type, so allow one level of optionals
```

Comment:
Maybe just mention this as an example and document the general case:

 ```suggestion
		// Reference expressions may reference a value which has an optional type.
		// For example, the result of indexing into a dictionary is an optional:
		//
		// let ints: {Int: String} = {0: "zero"}
		// let ref: &T? = &ints[0] as &T?   // read as (&T)?
		//
		// In this case the reference expression's type is an optional type.
		// Unwrap it one level to get the actual reference type
```